include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+
+  /**------------- Begin MRU helper method definitions -------------------**/
+  // perform a similiar work for cache.lines.constraint(cache.regs)
+  // This one requires to be renaming because mem base dom used different base
+  // vars
+  base_abstract_domain_t cache_lines_regs_reduce() {
+    base_abstract_domain_t res(m_cache_lines);
+    auto csts = m_regs.to_linear_constraint_system();
+    res += csts;
+    return res;
+  }
+
+  // get all used Crab IR vars by region dom
+  void vars(variable_vector_t &used_vars) const {
+    used_vars.reserve(m_var_map.size());
+    for (auto &kv : m_var_map) {
+      used_vars.push_back(kv.first);
+    }
+  }
+
+  void region_vars_in_mem(variable_vector_t &used_vars) const {
+    used_vars.reserve(m_var_map.size());
+    for (auto &kv : m_var_map) {
+      if (kv.first.get_type().is_region()) {
+        used_vars.push_back(kv.first);
+      }
+    }
+  }
+
+  void region_vars_in_cache(variable_vector_t &used_vars) const {
+    used_vars.reserve(m_cache_var_map.size());
+    for (auto &kv : m_cache_var_map) {
+      if (kv.first.get_type().is_region()) {
+        used_vars.push_back(kv.first);
+      }
+    }
+  }
+
+  void vars_in_cache(variable_vector_t &used_vars) const {
+    used_vars.reserve(m_cache_var_map.size());
+    for (auto &kv : m_cache_var_map) {
+      used_vars.push_back(kv.first);
+    }
+  }
+
+  variable_vector_t vec_minus(variable_vector_t left, variable_vector_t right) {
+    using variable_map_t = ikos::patricia_tree<variable_t, variable_t>;
+    variable_map_t vars;
+    for (auto it = left.begin(); it != left.end(); ++it) {
+      vars.insert(*it, *it);
+    }
+    for (auto it = right.begin(); it != right.end(); ++it) {
+      if (vars.find(*it) != nullptr)
+        vars.remove(*it);
+    }
+    variable_vector_t res;
+    for (auto it = vars.begin(); it != vars.end(); ++it) {
+      res.push_back(it->first);
+    }
+    return res;
+  }
+
+  // NOTE: For debug use
+  void print_base_vec(const base_variable_vector_t &vec) {
+    for (auto elem : vec)
+      crab::outs() << elem << " ";
+    crab::outs() << "\n";
+  }
+
+  // NOTE: For debug use
+  void print_var_map() {
+    crab::outs() << "Cache varmap\t\n";
+    for (auto &kv : m_var_map)
+      crab::outs() << kv.first << " ";
+    crab::outs() << "\n";
+  }
+
+  // NOTE: For debug use
+  void print_vec(const variable_vector_t &vec) {
+    for (auto elem : vec)
+      crab::outs() << elem << " ";
+    crab::outs() << "\n";
+  }
+
+  // TODO: need to implement this on next iteration
+  void invalidate_cache() {}
+
+  // fresh cache as: { cache lines: bot, regs: top, addrs: addrs -= C_base }
+  // for addrs, we keep all equalities of base address for each reference
+  // also need to assign cache used flag C_used as false
+  void make_fresh_cache() {
+    m_cache_lines.set_to_bottom();
+    m_regs.set_to_top();
+    m_addrs.set_to_top();
+    if (m_base) {
+      forget(*m_base);
+    }
+    if (m_used) {
```

Comment:
I understand now why you define `m_used` as a variable. I think this code is super complicated. Why not just a `boolean_value`?